src/backend/metal/src/command.rs

```diff
@@ -453,176 +147,239 @@ impl command::Buffer<Resources> for CommandBuffer {
-        self.index_buf = Some((buf, idx_type));
+        use map::map_index_type;
+
+        // TODO(fkaa): pass wrapper instead
+        self.encoder.set_index_buffer(unsafe { *(buf.0).0 }, map_index_type(idx_type));
     }
 
     fn set_scissor(&mut self, rect: target::Rect) {
-        self.buf.push(Command::SetScissor(MTLScissorRect {
+        // TODO(fkaa): why are getting 1x1 scissor?
+        /*self.encoder.set_scissor_rect(MTLScissorRect {
             x: rect.x as u64,
             y: rect.y as u64,
             width: (rect.x + rect.w) as u64,
             height: (rect.y + rect.h) as u64,
-        }));
+        });*/
     }
 
     fn set_ref_values(&mut self, vals: state::RefValues) {
-        if vals.stencil.0 != vals.stencil.1 {
-            error!("Unable to set different stencil ref values for front ({}) and back ({})",
-                   vals.stencil.0,
-                   vals.stencil.1);
-        }
+        // FIXME: wrong types?
+        // self.encoder.set_stencil_front_back_reference_value(vals.stencil.0, vals.stencil.1);
+
         // TODO: blend/stencil
     }
 
     fn update_buffer(&mut self, buf: Buffer, data: &[u8], offset: usize) {
-        if self.in_use.contains(&buf) {
-            unsafe {
-                if self.encoding {
-                    self.render_encoder.end_encoding();
+        use map::{map_buffer_usage};
 
-                    if self.root {
-                        (*self.master_encoder).end_encoding();
-                        // (*self.master_encoder).release();
-                        *self.master_encoder = MTLParallelRenderCommandEncoder::nil();
+        debug_assert!(!unsafe { *(*(buf.0).0) }.is_null(), "Buffer must be non-nil");
 
-                        (*self.mtl_buf).present_drawable(*self.drawable);
-                        (*self.mtl_buf).commit();
-                        (*self.mtl_buf).wait_until_completed();
-                        (*self.mtl_buf).release();
+        let b = unsafe { *(buf.0).0 };
 
-                        *self.master_encoder = MTLParallelRenderCommandEncoder::nil();
-                        *self.mtl_buf = MTLCommandBuffer::nil();
+        // we can create a new buffer to avoid synchronization when:
+        //   * buffer is write-only
+        //   * we are replacing all the contents of the buffer
+        //
+        // TODO(fkaa): maybe have an upper limit to prevent from thrashing?
+        if offset == 0 && data.len() == b.length() as usize {
+            unsafe {
+                // TODO(fkaa): ensure the creation flags are identical
+                *(buf.0).0 = self.device.new_buffer_with_data(
+                    data.as_ptr() as _,
+                    data.len() as _,
+                    map_buffer_usage(buf.1));
+
+                // invalidate old buffer in cache
+                self.encoder.invalidate_buffer(b);
+            }
+        } else {
+            // TODO(fkaa): how do we (eventually) handle updating buffers when
```

Comment:
good question. We need a solid synchronization model for resources (buffers and textures)